python/src/iceberg/io/boto.py

```diff
@@ -0,0 +1,294 @@
+
+    def __init__(self, location: str, session: Session):
+        parsed_location = urlparse(location)  # Create a ParseResult from the uri
+
+        if not parsed_location.scheme.startswith('s3'):
+            raise ValueError(f"Cannot create BotoOutputFile, scheme not supported: {parsed_location.scheme}")
+
+        self._bucket = parsed_location.netloc
+        self._path = parsed_location.path.strip("/")
+        self._session = session
+        super().__init__(location=location)
+
+    def __len__(self) -> int:
+        """Returns the total length of the file, in bytes"""
+        file_info = self._file_info()
+        return file_info.size
+
+    def exists(self) -> bool:
+        """Checks whether the location exists"""
+        try:
+            self._session.resource('s3').Bucket(self._bucket).Object(self._path).load()  # raises botocore.exceptions.ClientError with a 404 if it does not exist
+            return True
+        except ClientError as e:
+            if e.response['Error']['Code'] == "404":
+                return False
+            else:
+                raise
+
+    def create(self, overwrite: bool = False) -> OutputStream:
+        """Create a BotoOutputStream for reading the contents of the file
+
+        Args:
+            overwrite(bool): Whether to overwrite the file if it already exists
+
+        Returns:
+            BotoOutputStream: A NativeFile instance for the file located at self.location
+
+        Raises:
+            FileExistsError: If the file already exists at `self.location` and `overwrite` is False
+
+        Note:
+            If overwrite is set to False, a check is first performed to verify that the file does not exist.
+            This is not thread-safe and a possibility does exist that the file can be created by a concurrent
+            process after the existence check yet before the output stream is created. In such a case, the default
+            behavior will truncate the contents of the existing file when opening the output stream.
+        """
+        return BotoOutputStream(s3_object=self._session.resource('s3').Bucket(self._bucket).Object(self._path))
+
+    def to_input_file(self) -> BotoInputFile:
+        """Returns a new BotoInputFile for the location"""
+        return BotoInputFile(location=self.location)
+
+
+class BotoFileIO(FileIO):
+    """A FileIO implementation that uses boto3
+    
+    Example1:
+        >>> from iceberg.io.boto import BotoFileIO
+        >>> file_io = BotoFileIO()
+        >>> input_file = file_io.input_file("s3://foo/bar.txt")
+        >>> input_stream = input_file.open()
+        >>> data = input_stream.read()
+    
+    Example2:
+        >>> from iceberg.io.boto import BotoFileIO
+        >>> file_io = BotoFileIO()
+        >>> output_file = file_io.output_file("s3://foo/bar.txt")
+        >>> output_stream = output_file.create()
+        >>> output_stream.write(b'foo')
+    """
+    def __init__(self, session: Optional[Session] = None):
```

Comment:
My understanding of the intention was the by passing the optional session, you can let the user to whatever magic configuration they needed (specifically endpoint_url).

But when I read [the documentation here](https://boto3.amazonaws.com/v1/documentation/api/latest/reference/core/session.html), I get the impression that it's in fact the "client" or "resource" where this is configured, not the session.

maybe I misunderstand something?

Reply:
Ah you're absolutely right. But since we use the session object directly to create resource clients we'll probably need to add an endpoint argument to the `BotoFileIO` init and then consistently pass it to `resource(...)` which also has an [endpoint_url](https://boto3.amazonaws.com/v1/documentation/api/latest/reference/core/session.html#boto3.session.Session.resource) argument. Let me update this, thanks @joshuarobinson!

Reply:
maybe use a kwargs?

as a point of reference, this is how I initialize connections using s3fs:
`storage_options = {'endpoint_url': ENDPOINT_URL}
fs = fsspec.filesystem('s3', client_kwargs=storage_options)`

Reply:
Love it, I added a **client_kwargs argument and pass it through to the resource